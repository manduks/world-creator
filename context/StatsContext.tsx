import * as React from "react";
import { Coordinates } from 'types/world';

type StatsProviderProps = {children: React.ReactNode}
interface StatsInterface {
  cursor?: Coordinates;
  squaresCount?: number;
  islandsCount?: number;
}
interface Action extends StatsInterface { 
  type: 'cursor' | 'square' | 'reset'
}
type Dispatch = (action: Action) => void

const initialState = {
  cursor: {
    x: 0,
    y: 0,
  },
  squaresCount: 0,
  islandsCount: 0,
};
const StatsContext = React.createContext<StatsInterface | undefined>(undefined);
const StatsDispatchContext = React.createContext<Dispatch | undefined>(undefined);

function statsReducer(state: StatsInterface, action: Action) {
  switch (action.type) {
    case 'cursor': {
      return { ...state, cursor: action.cursor };
    }
    case 'square': {
      return { ...state, squaresCount: action.squaresCount, islandsCount:  action.islandsCount};
    }
    case 'reset': {
      return {
        ...initialState,
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}

function StatsProvider({ children }: StatsProviderProps) {
  const [state, dispatch] = React.useReducer(statsReducer, initialState);

  return (
    <StatsContext.Provider value={state}>
      <StatsDispatchContext.Provider value={dispatch}>
        {children}
      </StatsDispatchContext.Provider>
    </StatsContext.Provider>
  )
}

function useStatsContext() {
  const context = React.useContext(StatsContext);
  if (context === undefined) {
    throw new Error('useStatsContext must be used within a StatsProvider')
  }
  return context;
}

function useStatsDispatch() {
  const context = React.useContext(StatsDispatchContext);
  if (context === undefined) {
    throw new Error('useStatsDispatch must be used within a StatsProvider')
  }
  return context;
}

export { StatsProvider, useStatsContext, useStatsDispatch };

