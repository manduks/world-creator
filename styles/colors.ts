export const BLUE = "#3B82F6";
export const GREEN = "#10B981";

export const colors = [BLUE, GREEN];
