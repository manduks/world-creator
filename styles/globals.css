/* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
html,
body,
div,
span,
applet,
object,
iframe,
h1,
h2,
h3,
h4,
h5,
h6,
p,
blockquote,
pre,
a,
abbr,
acronym,
address,
big,
cite,
code,
del,
dfn,
em,
img,
ins,
kbd,
q,
s,
samp,
small,
strike,
strong,
sub,
sup,
tt,
var,
b,
u,
i,
center,
dl,
dt,
dd,
ol,
ul,
li,
fieldset,
form,
label,
legend,
table,
caption,
tbody,
tfoot,
thead,
tr,
th,
td,
article,
aside,
canvas,
details,
embed,
figure,
figcaption,
footer,
header,
hgroup,
menu,
nav,
output,
ruby,
section,
summary,
time,
mark,
audio,
video {
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}
/* HTML5 display-role reset for older browsers */
article,
aside,
details,
figcaption,
figure,
footer,
header,
hgroup,
menu,
nav,
section {
  display: block;
}
body {
  line-height: 1;
}
ol,
ul {
  list-style: none;
}
blockquote,
q {
  quotes: none;
}
blockquote:before,
blockquote:after,
q:before,
q:after {
  content: "";
  content: none;
}
table {
  border-collapse: collapse;
  border-spacing: 0;
}

html,
body {
  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
    Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
}

.container {
  height: 100vh;
  width: 100vw;
}

* {
  box-sizing: border-box;
}

main {
  height: 100vh;
  width: 100vw;
  color: #fefefe;
  background-color: #010312;
}

h1 {
  font-size: 2em;
  text-align: center;
}

.header {
  height: 100px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.header span {
  padding: 12px;
}

.world-container {
  display: block;
  border: 1px solid blue;
  width: 100%;
  height: calc(100vh - 160px); /* header and footer */
}

.cell:hover {
  border: 2px solid #6ee7b7;
}

footer {
  height: 60px;
  position: fixed;
  width: 100%;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

/*inputs*/
.footer-inputs {
  display: flex;
  flex-direction: row;
}

.input-container {
  display: flex;
  flex-direction: row;
  width: 200px;
  height: 40px;
  align-items: center;
  font-size: 24px;
}

.input-container input {
  height: 40px;
  width: 100px;
  align-items: center;
  justify-content: center;
  background-color: #010312;
  border: 2px solid #374151;
  border-radius: 4px;
  color: white;
  font-size: 24px;
  text-align: center;
}
.input-container label {
  padding: 0 8px;
}

/*cursor*/
.cursor {
  display: flex;
  flex-direction: column;
  font-size: 12;
  padding: 16px;
  width: 120px;
}
.cursor span {
  padding: 2px;
}

/*stats*/
.stats {
  display: flex;
  flex-direction: row;
  font-size: 12;
  padding: 16px;
  font-size: 18px;
}
.stats div {
  display: flex;
  flex-direction: row;
  padding: 8px;
  align-items: center;
}
.square {
  display: block;
  width: 20px;
  height: 20px;
  border: 1px solid #10b981;
  margin: 0 8px;
}
.island {
  background-color: #10b981;
}
