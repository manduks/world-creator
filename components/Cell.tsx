import * as React from 'react';
import { colors } from 'styles/colors';
import { getWorld, calculateIslands, setLand, deleteLand } from 'lib/world';
import { useStatsDispatch } from 'context/StatsContext';

interface CellProps {
  columnIndex: number;
  rowIndex: number;
  style: object;
}

export default function Cell({ columnIndex, rowIndex, style }: CellProps) {
  const dispatch = useStatsDispatch();
  const memPosition = `${columnIndex}_${rowIndex}`;
  const world = getWorld();
  const [color, setColor] = React.useState(world[memPosition] || 0);

  // debugger;
  // this is needed by react-virtualized to render the cell correctly
  style = {
    ...style,
    backgroundColor: colors[color],
  };

  return (
    <div
      className="cell"
      data-testid="grid-cell"
      style={style}
      onClick={() => {
        const newValue = 1 - color; // toggle color between 0 and 1
        debugger;
        if (newValue) {
          setLand(memPosition, newValue);
        } else {
          // clean up memory for easier loop later
          deleteLand(memPosition);
        }
        setColor(newValue);
        dispatch({
          type: 'square',
          islandsCount: calculateIslands(),
          squaresCount: Object.keys(world).length,
        });
      }}
      onMouseEnter={() => {
        dispatch({ type: 'cursor', cursor: { x: columnIndex, y: rowIndex } });
      }}
    ></div>
  );
}
