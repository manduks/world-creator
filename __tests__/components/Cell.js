import * as React from 'react'
import renderer from 'react-test-renderer';
import { render, fireEvent, screen } from '@testing-library/react';
import Cell from 'components/Cell'
import { StatsProvider } from 'context/StatsContext';
import { GREEN, BLUE } from 'styles/colors';

describe('<Cell />', () => {
  it('renders component', () => {
    const tree = renderer.create(
      <StatsProvider>
        <Cell />    
      </StatsProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  });

  it('renders component with context and verifies click event', () => {
    render(
      <StatsProvider>
        <Cell />    
      </StatsProvider>
    );
    fireEvent.click(screen.getByTestId('grid-cell'));
    expect(screen.getByTestId('grid-cell')).toHaveStyle(`background-color: ${GREEN}`);
    fireEvent.click(screen.getByTestId('grid-cell'));
    expect(screen.getByTestId('grid-cell')).toHaveStyle(`background-color: ${BLUE}`);
  });
});
