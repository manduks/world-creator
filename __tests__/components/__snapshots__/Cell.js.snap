// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> renders component 1`] = `
<div
  className="cell"
  data-testid="grid-cell"
  onClick={[Function]}
  onMouseEnter={[Function]}
  style={
    Object {
      "backgroundColor": "#3B82F6",
    }
  }
/>
`;
